### PR TITLE
Add domains support

### PR DESCRIPTION
**Pull request by gerrod3**
Still working on this, current behavior is reliant on docs endpoint being still exposed at original endpoint `/pulp/api/v3/docs/api.json`. 

Review Checklist:
- [ ] An issue is properly linked. [feature and bugfix only]
- [ ] Tests are present or not feasible.
- [ ] Commits are split in a logical way (not historically).

### CODE REVIEW

File: pulpcore/cli/common/openapi.py [hunk 1]
```diff
@@ -291,6 +293,11 @@ def call(
             parameters = {}
         else:
             parameters = parameters.copy()
+        if self.api_spec["info"].get("x-pulp-domain-enabled", False):
```
Comment (mdellweg) [inline pulpcore/cli/common/openapi.py:296]: I know, this is an early draft, but i'd like to keep any thing Pulp-specific out of the `openapi.py` file. Can this be handled in `context.py`?

Comment (gerrod3) [reply pulpcore/cli/common/openapi.py]: I put the domains check into `context.py`, but I had to add a new method to `OpenApi`, `param_spec(operation_id, param_type)`, so I could know when to add `domains_path` to the parameters.

File: tests/scripts/pulp_file/test_acs.sh [hunk 2]
```diff
@@ -38,7 +38,7 @@ test "$(echo "$OUTPUT" | jq ".paths | length")" -eq 2
 
 # test refresh
 expect_succ pulp --background file acs refresh --acs $acs
-task_group=$(echo "$ERROUTPUT" | grep -E -o "${PULP_API_ROOT}api/v3/task-groups/[-[:xdigit:]]*/")
+task_group=$(echo "$ERROUTPUT" | grep -E -o "/[/a-zA-Z0-9_]*/api/v3/task-groups/[-[:xdigit:]]*/")
```
Comment (ggainey) [inline tests/scripts/pulp_file/test_acs.sh:41]: This regex will allow some really odd URLs. 

PULP_API_ROOT starts and ends with /, domain-name is "just" the name. Can we put an if-domain switch here, with something like 

```suggestion
task_group=$(echo "$ERROUTPUT" | grep -E -o "${PULP_API_ROOT}{$domain_name}/api/v3/task-groups/[-[:xdigit:]]*/")
```

if-domain, and the existing check if-not-domain?

**Review by ggainey — CHANGES_REQUESTED** [review]

File: pulp-glue/pulp_glue/core/context.py [hunk 3]
```diff
@@ -87,6 +87,13 @@ def upload(
         return task["created_resources"][0]
 
 
+class PulpDomainContext(PulpEntityContext):
+    ENTITY = _("Pulp domain")
+    ENTITIES = _("Pulp domains")
+    HREF = "domain_href"
+    ID_PREFIX = "domains"
```
Comment (mdellweg) [inline pulp-glue/pulp_glue/core/context.py:94]: This should have a NEED_PLUGINS.

File: tests/scripts/pulp_file/test_acs.sh (hunk 2 above)
Comment (mdellweg) [reply tests/scripts/pulp_file/test_acs.sh:41]: It is also de facto of a subtle check that the href has the expected format. I think the grep would fail our test. I do not feel stong either way. If you simplify, however we can just use `.*`.

File: pulpcore/cli/common/generic.py [hunk 4]
```diff
@@ -490,7 +490,11 @@ def _option_callback(
 
         if value.startswith("/"):
             # The HREF of a resource was passed
-            pattern = rf"^{pulp_ctx.api_path}{entity_ctx.HREF_PATTERN}"
+            href_pattern = entity_ctx.HREF_PATTERN
+            if pulp_ctx.domain_enabled:
+                pattern = rf"^{pulp_ctx._api_root}(?P<pulp_domain>\w+)/api/v3/{href_pattern}"
```
Comment (mdellweg) [inline pulpcore/cli/common/generic.py:495]: Why is this needed? We only accept entities of the proper domain anyway. If the domain in hrefs is optional (I don't think it is), we can still do `rf"^{pulp_ctx._api_root}(?{pulp_ctx.domain_name}/)api/v3/{href_pattern}"` right?

File: pulpcore/cli/common/generic.py [hunk 5]
```diff
@@ -547,7 +551,10 @@ def _option_callback(
                         option_name=param.name
                     )
                 )
-            pattern = rf"^{pulp_ctx.api_path}{href_pattern}"
+            if pulp_ctx.domain_enabled:
+                pattern = rf"^{pulp_ctx._api_root}(?P<pulp_domain>\w+)/api/v3/{href_pattern}"
+            else:
+                pattern = rf"^{pulp_ctx.api_path}{href_pattern}"
```
Comment (mdellweg) [inline pulpcore/cli/common/generic.py:557]: Same here.

File: pulpcore/cli/core/group.py [hunk 6]
```diff
@@ -193,13 +207,13 @@ def role(ctx: click.Context, pulp_ctx: PulpCLIContext, group_ctx: PulpGroupConte
         decorators=[
             group_option,
             click.option("--role", required=True),
-            click.option("--object", "content_object", required=True),
```
Comment (mdellweg) [inline pulpcore/cli/core/group.py:196]: This should not have been there in the first place, right?

File: pulpcore/cli/core/group.py [hunk 7]
```diff
@@ -185,6 +198,7 @@ def role(ctx: click.Context, pulp_ctx: PulpCLIContext, group_ctx: PulpGroupConte
                 callback=null_callback,
                 help=_('Filter roles by the associated object. Use "" to list global assignments.'),
             ),
+            domain_option,
```
Comment (mdellweg) [inline pulpcore/cli/core/group.py:201]: Should we add a guard that certain commands can only be used in the default domain?

File: pulpcore/cli/core/group.py [hunk 8]
```diff
@@ -193,13 +207,13 @@ def role(ctx: click.Context, pulp_ctx: PulpCLIContext, group_ctx: PulpGroupConte
         decorators=[
             group_option,
             click.option("--role", required=True),
-            click.option("--object", "content_object", required=True),
             click.option(
                 "--object",
                 "content_object",
-                required=True,
+                default="",
```
Comment (mdellweg) [inline pulpcore/cli/core/group.py:213]: It is required on purpose. Assigning global permissions should not be done by accident.

File: pulpcore/cli/core/group.py [hunk 9]
```diff
@@ -217,11 +231,19 @@ def role(ctx: click.Context, pulp_ctx: PulpCLIContext, group_ctx: PulpGroupConte
...
                 callback=lookup_callback("content_object", PulpGroupRoleContext),
                 expose_value=False,
                 help=_('Associated object; use "" for global assignments.'),
             ),
+            pulp_option(
+                "--domain",
+                # This only accepts HREFs
+                callback=lookup_callback("domain", PulpGroupRoleContext),
```
Comment (mdellweg) [inline pulpcore/cli/core/group.py:242]:
```suggestion
                callback=lookup_callback("domain", PulpDomainContext),
```
does this help to revisit you comment about HREFs?

File: pulp-glue/pulp_glue/common/context.py [hunk 10]
```diff
@@ -90,15 +90,18 @@ def __init__(
...
         api_kwargs: Dict[str, Any],
+        domain: Optional[str],
         format: str,
         background_tasks: bool,
         timeout: int,
     ) -> None:
         self._api: Optional[OpenAPI] = None
-        self.api_path: str = api_root + "api/v3/"
+        self._api_root: str = api_root
+        self._normal_api_path: str = api_root + "api/v3/"
```
[the remark this answers is not in the document]
Comment (gerrod3) [reply pulp-glue/pulp_glue/common/context.py:100]: Yes, the status and docs endpoints are both still available at the normal paths.

File: pulp-glue/pulp_glue/common/context.py [hunk 11]
```diff
@@ -153,13 +156,29 @@ def _patch_api_spec(self) -> None:
...
+
     @property
     def api(self) -> OpenAPI:
         if self._api is None:
             if self._api_kwargs.get("username") and not self._api_kwargs.get("password"):
                 self._api_kwargs["password"] = self.prompt("password", hide_input=True)
             try:
-                self._api = OpenAPI(doc_path=f"{self.api_path}docs/api.json", **self._api_kwargs)
+                self._api = OpenAPI(
+                    doc_path=f"{self._normal_api_path}docs/api.json", **self._api_kwargs
```
[the remark this answers is not in the document]
Comment (gerrod3) [reply pulp-glue/pulp_glue/common/context.py:180]: Yep, api-docs will also be available in each domain.

File: pulp-glue/pulp_glue/common/context.py [hunk 12]
```diff
@@ -188,8 +207,12 @@ def call(
         Returns the operation result, or the finished task.
         If non_blocking, returns unfinished tasks.
         """
-        if parameters is not None:
-            parameters = preprocess_payload(parameters)
+        if parameters is None:
+            parameters = {}
+        if self.domain_enabled:
+            if "pulp_domain" in self.api.param_spec(operation_id, "path", required=True):
+                parameters["pulp_domain"] = self.pulp_domain
```
[the remark this answers is not in the document]
Comment (gerrod3) [reply pulp-glue/pulp_glue/common/context.py:214]: This is correct, detail endpoints don't require the pulp_domain path parameter (since they already have it filled in) so we can't blindly add it to each call.

File: pulp-glue/pulp_glue/common/openapi.py [hunk 13]
```diff
@@ -150,6 +150,29 @@ def _set_correlation_id(self, correlation_id: str) -> None:
         else:
             self._session.headers["Correlation-ID"] = correlation_id
 
+    def param_spec(
```
[the remark this answers is not in the document]
Comment (gerrod3) [reply pulp-glue/pulp_glue/common/openapi.py:153]: Probably, I didn't want to touch the openapi code more than I had to. Maybe in a future refactor.

File: pulpcore/cli/common/generic.py (hunk 4 above)
Comment (gerrod3) [reply pulpcore/cli/common/generic.py:495]: This is for user/group role assignment. Assigning an object-level role can have the object be in a different domain than the user/group which are always in the default domain.

File: pulpcore/cli/core/group.py (hunk 7 above)
Comment (gerrod3) [reply pulpcore/cli/core/group.py:201]: Possibly, currently Pulp should disallow this, but having a nice error message from the CLI would be helpful.

File: pulpcore/cli/core/group.py (hunk 6 above)
Comment (gerrod3) [reply pulpcore/cli/core/group.py:196]: Correct, it was  duplicate option.

File: pulpcore/cli/core/group.py (hunk 8 above)
Comment (gerrod3) [reply pulpcore/cli/core/group.py:213]: Added a callback to keep current behavior if domain is not specified.

File: pulpcore/cli/core/group.py (hunk 9 above)
Comment (gerrod3) [reply pulpcore/cli/core/group.py:242]: No, the `lookup_callback` only adds the field to the entity context's lookup and requires that the context exists in the command, which `PulpDomainContext` does not.

 I added more functionality to `resource_option` to allow for specifying the domain name for this command.

File: pulp-glue/pulp_glue/common/context.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mdellweg) [reply pulp-glue/pulp_glue/common/context.py]: But still we need to assemble the string manually, because the `api_root` property would call to the server giving us an infinite loop. Also when using the non-domained api docs, we only save a local copy once (pretty much preferred).

File: pulpcore/cli/core/group.py [hunk 14]
```diff
@@ -217,11 +247,12 @@ def role(ctx: click.Context, pulp_ctx: PulpCLIContext, group_ctx: PulpGroupConte
             click.option(
                 "--object",
                 "content_object",
-                required=True,
+                default="",
```
Comment (mdellweg) [inline pulpcore/cli/core/group.py:250]:
```suggestion
                callback=_object_required_callback,
```

File: pulpcore/cli/core/user.py [hunk 15]
```diff
@@ -124,11 +154,12 @@ def role(ctx: click.Context, pulp_ctx: PulpCLIContext, user_ctx: PulpUserContext
             click.option(
                 "--object",
                 "content_object",
-                required=True,
+                default="",
```
Comment (mdellweg) [inline pulpcore/cli/core/user.py:157]:
```suggestion
                callback=_object_required_callback,
```

File: tests/scripts/pulp_file/test_acs.sh (hunk 2 above)
Comment (mdellweg) [reply tests/scripts/pulp_file/test_acs.sh:41]:
```suggestion
task_group=$(echo "$ERROUTPUT" | grep -E -o "${PULP_API_ROOT}\(/[_a-zA-Z0-9]*\)?/api/v3/task-groups/[-[:xdigit:]]*/")
```

**Review by mdellweg — COMMENTED** [review]
I think you missed two places.

File: tests/scripts/pulpcore/test_pulpimporter.sh [hunk 16]
```diff
@@ -5,6 +5,8 @@
 
 pulp debug has-plugin --name "file" || exit 23
 
+eval "$(pulp status | jq -r '.domain_enabled // false | not')" || exit 23
```
Comment (mdellweg) [inline tests/scripts/pulpcore/test_pulpimporter.sh:8]: We could have a `pulp debug` command like `pulp debug has-plugin`.

**Review by mdellweg — APPROVED** [review]

File: tests/scripts/pulpcore/test_pulpimporter.sh (hunk 16 above)
Comment (ggainey) [reply tests/scripts/pulpcore/test_pulpimporter.sh:8]: > We could have a `pulp debug` command like `pulp debug has-plugin`.

@mdellweg  Open an issue pliz!

**Review by ggainey — APPROVED** [review]
OK, I think we are good to go - thanks for all the work gerrod!
